build bind chroot when service starts

hrosenfeld opened the pull request:
This change will greatly improve support for running bind in a chroot. Minimal support for chroot was there already, but that required manually maintaining (and updating) the chroot, which I forgot every time I updated my system and every now and then it wouldn't come up cleanly because of this.

The dns-server method will now populate the chroot with everything needed. It will require additional privileges, but it will drop them when starting named. Additionally it will support running named as a non-root user.

hrosenfeld committed:
Build bind chroot when service starts.

pyhalov commented on components/bind/Solaris/dns-server.sh:
```diff
@@ -99,12 +100,54 @@ case "$method" in
+
     # If chroot option is set, note zones(5) are preferred, then
     # configuration file lives under chroot directory.
     if [ "${chroot_dir}" != "" ]; then
```
I like this, but please, add also check for $chroot_dir != / , to disallow user to shoot himself in the foot.

pyhalov commented:
Also, please bump COMPONENT_REVISION in Makefile

hrosenfeld commented:
Added the requested check and set COMPONENT_REVISION in Makefile.

I also changed the way the configuration files are propagated to the chroot. I now create symlinks to the chrooted files in the original location so that a sysadmin can edit the files in their regular location and rndc still works.

pyhalov commented on components/bind/Solaris/dns-server.sh:
```diff
@@ -99,12 +100,67 @@ case "$method" in
+
+      if [ ! -L ${configuration_file} ]; then
+        pax -rw -H -pe ${configuration_file} ${chroot_dir}
+        ln -sf ${chroot_dir}/${configuration_file} ${configuration_file}
```
Isn't it safer to mount configuration files to chroot dir, without replacing them with links?

hrosenfeld commented:
All config files and the pkcs11 libs will be lofs-mounted into the chroot. Unless you have further change requests I'd consider this ready for merge.

pyhalov commented on components/bind/Solaris/dns-server.sh:
```diff
@@ -20,34 +20,39 @@
-fi
-server="/usr/sbin/named"
-I=`/usr/bin/basename $0`
+        umount ${c}/${f}
```
Shouldn't we check that file is mounted before unmounting it or at least suppress error messages?

hrosenfeld replied:
This is just to make sure it actually is not mounted already before mounting it. Under normal operation that is noop, so I chose to supress error messages.

pyhalov commented on components/bind/Solaris/dns-server.sh:
```diff
@@ -99,40 +104,97 @@ case "$method" in
+
+        server="env LD_NOLAZYLOAD=1 ${server}"
+
+        mkdir -p ${chroot_dir}
```
What if chroot_dir already exists?

hrosenfeld replied:
It's not an error for mkdir -p if a directory already exists, it will leave it alone.
